### DISCUSSION TITLE
Serving a SPA along side an API

### DISCUSSION THREAD

**Discussion by clarkmcc** [tokio-rs/axum#1472] [Q&A] [answered]
My design requirements are
* Serving a single-page-application that is embedded into the binary using rust-embed (I've followed [this example](https://github.com/pyrossh/rust-embed/blob/master/examples/axum.rs)) from the `/` path (root).
* Serve an API at the `/api/v1` path.

I've tried configuring my router like this
```rust
let app = Router::new()
        .nest(
            "/api/v1",
            Router::new().nest(
                "/auth",
                Router::new().route("/login", post(auth::login)),
            ),
        )
        .route("/*rest", get(gui::static_handler));
```

but the `/*rest` route is giving me this error.

```
thread 'main' panicked at 'Invalid route: insertion failed due to conflict with previously registered route: /api/v1/auth/login'
```

Is it not possible for me to register a set of API routes and then anything that's not matched get's handled by the gui handler? The gui handler does work as expected when the API routes are disabled.

**Comment by davidpdrsn** [comment, marked as answer]
Use `Router::fallback` instead.
